src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java

```diff
@@ -0,0 +1,252 @@
+      return deleteMetricsByMetricName(tenant, metricName, start, end);
+    } else {
+      return deleteMetricsByMetricNameAndTag(tenant, metricName, tag, start, end);
+    }
+  }
+
+  private Mono<Empty> deleteMetricsByTenantId(String tenant, Instant start, Instant end) {
+    log.debug("Deleting metrics for tenant: {}", tenant);
+    return Flux.fromIterable(timeSlotPartitioner.partitionsOverRange(start, end, null))
+        .flatMap(timeSlot ->
+            deleteMetricsByTenantId(downsampleProperties.getGranularities(), tenant, timeSlot))
+        .then(Mono.empty());
+  }
+
+  private Mono<Empty> deleteMetricsByMetricName(String tenant, String metricName, Instant start,
+      Instant end) {
+    log.debug("Deleting metrics {} for tenant: {} ", metricName,
+        tenant);
+    return Flux.fromIterable(timeSlotPartitioner.partitionsOverRange(start, end, null))
+        .flatMap(timeSlot -> {
+          Flux<String> seriesSetHashes = getSeriesSetHashFromSeriesSets(tenant,
+              metricName);
+          return deleteMetric(downsampleProperties.getGranularities(), tenant, timeSlot, metricName,
+              seriesSetHashes);
+        }).then(Mono.empty());
+  }
+
+  private Mono<Empty> deleteMetricsByMetricNameAndTag(String tenant, String metricName,
+      List<String> tag, Instant start, Instant end) {
+    log.debug(
+        "Deleting metrics {} with tag {} for tenant: {}  ",
+        metricName, tag, tenant);
+    Map<String, String> queryTags = convertPairsListToMap(tag);
+    return Flux.fromIterable(timeSlotPartitioner.partitionsOverRange(start, end, null))
+        .flatMap(timeSlot -> {
+          Flux<String> seriesSetHashes = metadataService.locateSeriesSetHashes(tenant,
+              metricName, queryTags);
+          return deleteMetric(downsampleProperties.getGranularities(), tenant, timeSlot, metricName,
+              seriesSetHashes);
+        }).then(Mono.empty());
+  }
+
+  private Mono<Boolean> deleteMetric(List<Granularity> granularities, String tenant,
+      Instant timeSlot, String metricName, Flux<String> seriesSetHashes) {
+    return seriesSetHashes.flatMap(
+        seriesSetHash -> Flux.fromIterable(granularities)
+            //delete entries from downsampled tables
+            .flatMap(granularity ->
+                deleteRawOrDownsampledEntries(
+                    dataTablesStatements.downsampleDeleteWithSeriesSetHash(granularity.getWidth()),
+                    tenant, timeSlot, seriesSetHash))
+            .then(Mono.just(true))
+    )
+        .then(Mono.just(true))
+        //delete entries from raw table
+        .flatMap(item ->
+            seriesSetHashes.flatMap(seriesSetHash ->
+                deleteRawOrDownsampledEntries(dataTablesStatements.getRawDeleteWithSeriesSetHash(),
+                    tenant, timeSlot, seriesSetHash))
+                .then(Mono.just(true))
+        )
+        //delete entries from metric_names table
+        .flatMap(result ->
```

Comment:
What if we have a metric with same metric name in database which is not deleted? 
This will cause inconsistency in data then I guess

Reply:
not deleted because of any issues with DB server? in that case, we are doing retry.